### ISSUE TITLE
Radio Driver Wrapper

### ISSUE THREAD

**Issue by GrahamDrive** [UMSATS/cdh-tsat#30]
# Radio Driver Wrapper

## Problem statement
The current radio drivers are basic and not designed for an RTOS.
We need to develop a wrapper for the radio driver to facilitate its integration with FreeRTOS.

To achieve this, we will need to conduct research and development to determine the best method for integrating the radio with various RTOS tasks.

## Desired Operation
The radio should operate independently of any specific task, ensuring the radio code is non-blocking. This is crucial because the radio involves extended waiting periods, and we want to avoid CPU hogging during high communication traffic.

## Possible Solution
One possible approach is to create a set of RTOS tasks dedicated to managing the radio resource. In this method, any task that wants to send data over the radio will enqueue its information, which will then be transmitted by the radio task.

**Comment by ArnavG-it** [comment]
## Progress so far

### RTOS task structure:

- **[RadioHandler](https://github.com/UMSATS/cdh-tsat/blob/10072b664b242e403be6c127a689db1e9af6a7f3/cdh-tsat-stm32project-nucleo/Core/Src/main.c#L872)**
  - Manages the radio's state and the TX/RX tasks. This involves unblocking those tasks and putting the radio in transmission or reception mode when needed.

- **[RadioTxTask](https://github.com/UMSATS/cdh-tsat/blob/10072b664b242e403be6c127a689db1e9af6a7f3/cdh-tsat-stm32project-nucleo/Core/Src/main.c#L801)**
  -  Moves data from the `radioTxQueue` to the radio’s TX FIFO.

- **[RadioRxTask](https://github.com/UMSATS/cdh-tsat/blob/10072b664b242e403be6c127a689db1e9af6a7f3/cdh-tsat-stm32project-nucleo/Core/Src/main.c#L844)**
  -  Moves data from the radio’s RX FIFO into the `radioRxQueue`.

- **[RadioISR](https://github.com/UMSATS/cdh-tsat/blob/10072b664b242e403be6c127a689db1e9af6a7f3/cdh-tsat-stm32project-nucleo/Core/Src/main.c#L941)**
  -  Handles interrupts from the radio. It reads the interrupt status registers and notifies `RadioHandler` when a transmission or reception has been completed.

This structure isn't final and probably has room for improvement. For example, maybe combining `RadioTxTask` and `RadioRxTask`, or merging them into `RadioHandler`, would be better. Maybe having a separate task to read interrupt status registers after `RadioISR` receives an interrupt would minimize the time spent in an ISR. Maybe the handler could be a state machine that makes use of the [S2LP_STATE enum](https://github.com/UMSATS/cdh-tsat/blob/10072b664b242e403be6c127a689db1e9af6a7f3/cdh-tsat-stm32project-nucleo/Drivers/Hardware_Peripherals/Inc/SP-L2_driver.h#L51). Things like that.

### Current state:

The tasks are in their most minimal state, just enough for testing. Graham and I tried testing this a couple times, but were running into connection issues with the radio and board. Graham is working on testing this code on his own computer, since the radio drivers were working successfully there.

### Things to work on:

- **TX and RX queue management**: This involves fleshing out `RadioTxTask` and `RadioRxTask` to reliably move data between the queues and FIFOs. There may be some extra considerations with the queues. For example, if a TX message takes up multiple spots in the queue, its pieces should be next to each other and it may need a size indicator so that `RadioTxTask` can make sure the entire message is being moved into the FIFO without being split up.
 
- **Timeout and error handling**: We may need to account for tasks timing out or the radio entering an unexpected state.

- **Race conditions**: Ensuring that the tasks are properly synchronized via thread flags and that they aren't fighting for control over the radio. This could also mean making sure the FIFOs aren't being modified while the radio isn't in its READY state, and any other possible conditions.

- **Conditions for TX/RX**: `RadioHandler` needs to know when it should tell the radio to start transmitting or receiving. I'm not sure what this would entail.

I'll update this comment if I think of anything else.

### S2_LP datasheet (because it can be hard to find sometimes): https://www.st.com/resource/en/datasheet/s2-lp.pdf
